dash/backend/src/modules/reports/dao/reports.dao.ts

```diff
@@ -531,113 +531,94 @@ export class ReportsDao {
+      const imageDetailsQuery = knex.select([
+        'i.id',
+        knex.raw("concat(i.name,':', i.tag) as image"),
+        'i.running_in_cluster',
+        knex.raw("array_agg(distinct hki.namespace) as namespaces"),
+        knex.raw("array_agg(distinct hki.saved_date) as saved_dates")
+      ])
+        .from('images as i')
+        .leftJoin(knex.raw('clusters as c on c.id=i.cluster_id'))
+        .innerJoin(knex.raw('history_kubernetes_images as hki on hki.image_id = i.id'))
+        .where('i.deleted_at', null)
+        .andWhere('c.deleted_at', null)
+        .whereIn('hki.saved_date', [startDate, endDate]);
 
-        const newVulns = await knex.raw(rawQuery,
-            limit ? [endQuery, startQuery, limit] : [endQuery, startQuery])
-            .then(response => {
-            return plainToClass(ReportsVulnerabilityExportDto, response.rows);
+      if (clusterId) {
+        imageDetailsQuery.andWhere('hki.cluster_id', clusterId);
+      }
+      if (namespaces) {
+        imageDetailsQuery.whereIn('hki.namespace', namespaces);
+      }
+      imageDetailsQuery.groupBy(['i.id', 'i.name', 'hki.name']);
+
+      const scanResultsQuery = knex.select([
+        "image_details.saved_dates",
+        "isrs.scanner_name",
+        knex.raw('image_details.id as image_id'),
+        "image_details.image",
+        "image_details.running_in_cluster",
+        "isrs.type",
+        "isrs.severity",
+        "isrs.is_fixable",
+        "image_details.namespaces"
+      ])
+        .from(imageDetailsQuery.as('image_details'))
+        .leftJoin(knex.raw('image_scan_results as isr on isr.image_id = image_details.id and isr.is_latest=true'))
+        .innerJoin(knex.raw('image_scan_results_issues as isrs on isrs.image_results_id = isr.id'));
+
+      if (severityLevels) {
+          if (severityLevels.includes(VulnerabilitySeverity.MAJOR)) {
+              /** Major vulnerabilities are stored as 'High' on this database, ensures that the proper values
+               * are being searched for */
+              severityLevels.push(VulnerabilitySeverity.HIGH);
+          }
+        scanResultsQuery.whereIn('isrs.severity', severityLevels);
+      }
+      if (fixAvailable) {
+          scanResultsQuery.andWhere('isrs.is_fixable', fixAvailable);
+      }
+      const endDateAsDate = parseISO(endDate);
+      const validatedEndDate = endDateAsDate.toISOString().split('T')[0];
+      const startDateAsDate = parseISO(startDate);
+      const validatedStartDate = startDateAsDate.toISOString().split('T')[0];
+
+      const listsQuery = knex
+        .select(['scanner_name', 'image_id', 'image', 'running_in_cluster', 'type', 'severity', 'is_fixable', 'namespaces'])
+        .from(scanResultsQuery.as('scan_results'))
+        .whereRaw(knex.raw(`NOT saved_dates @> '{"${validatedStartDate}", "${validatedEndDate}"}'`));
+      const newVulnerabilities= listsQuery.clone().whereRaw(knex.raw(`NOT saved_dates @> '{"${validatedStartDate}"}'`));
```

Comment:
Starting here, you interpolate several values into knex.raw()'s.

I see that in this case those values are being generated so there shouldn't be a risk of SQL injection. It would still be a good idea to either add a comment making that clear, or to parameterize the knex.raw()'s

Parameterization example:
query.where(knex.raw('my_col = ? OR other_col = ?', [ val1, val2]));

Reply:
In this case, parameterization does not work. I tried a bunch of times haha. The quotes have to be exactly right and parameterization makes them unhappy.

Reply:
Added comments.